pkg/util/k8scmcache.go

```diff
@@ -0,0 +1,198 @@
+	cmDataKey = "content"
+
+	csiMetadataLabelAttr = "com.ceph.ceph-csi/persister"
+)
+
+func GetK8sNamespace() string {
+	namespace := os.Getenv("POD_NAMESPACE")
+	if namespace == "" {
+		return defaultNamespace
+	}
+	return namespace
+}
+
+func NewK8sClient() *k8s.Clientset {
+	var cfg *rest.Config
+	var err error
+	cPath := os.Getenv("KUBERNETES_CONFIG_PATH")
+	if cPath != "" {
+		cfg, err = clientcmd.BuildConfigFromFlags("", cPath)
+		if err != nil {
+			glog.Errorf("Failed to get cluster config with error: %v\n", err)
+			os.Exit(1)
+		}
+	} else {
+		cfg, err = rest.InClusterConfig()
+		if err != nil {
+			glog.Errorf("Failed to get cluster config with error: %v\n", err)
+			os.Exit(1)
+		}
+	}
+	client, err := k8s.NewForConfig(cfg)
+	if err != nil {
+		glog.Errorf("Failed to create client with error: %v\n", err)
+		os.Exit(1)
+	}
+	return client
+}
+
+func (k8scm *K8sCMCache) createMetadataCM(resourceID string) (*v1.ConfigMap, error) {
+	cm := &v1.ConfigMap{}
+	cm, err := k8scm.getMetadataCM(resourceID)
+	if err != nil {
+		glog.V(4).Infof("k8s-cm-cache: couldn't get configmap %s with error: %v", resourceID, err)
+		cm = &v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      resourceID,
+				Namespace: k8scm.Namespace,
+				Labels: map[string]string{
+					csiMetadataLabelAttr: cmLabel,
+				},
+			},
+			Data: map[string]string{},
+		}
+		cm, err = k8scm.Client.CoreV1().ConfigMaps(k8scm.Namespace).Create(cm)
+		if err != nil {
+			return nil, errors.Wrapf(err, "k8s-cm-cache: couldn't create configmap %s\n", resourceID)
+		}
+	}
+	glog.V(4).Infof("k8s-cm-cache: configmap %s successfully created\n", resourceID)
+	return cm, nil
+}
+
+func (k8scm *K8sCMCache) getMetadataCM(resourceID string) (*v1.ConfigMap, error) {
+	cm, err := k8scm.Client.CoreV1().ConfigMaps(k8scm.Namespace).Get(resourceID, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return cm, nil
+}
+
+func (k8scm *K8sCMCache) ForAll(pattern string, destObj interface{}, f ForAllFunc) error {
+	listOpts := metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", csiMetadataLabelAttr, cmLabel)}
+	cms, err := k8scm.Client.CoreV1().ConfigMaps(k8scm.Namespace).List(listOpts)
+	if err != nil {
+		return errors.Wrap(err, "k8s-cm-cache: failed to list metadata configmaps")
+	}
+
+	for _, cm := range cms.Items {
+		data := cm.Data[cmDataKey]
+		match, err := regexp.MatchString(pattern, cm.ObjectMeta.Name)
+		if err != nil {
+			continue
+		}
+		if !match {
+			continue
+		}
+		if err := json.Unmarshal([]byte(data), destObj); err != nil {
+			return errors.Wrap(err, "k8s-cm-cache: unmarshal error")
+		}
+		if err = f(cm.ObjectMeta.Name); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (k8scm *K8sCMCache) Create(identifier string, data interface{}) error {
+	cm, err := k8scm.getMetadataCM(identifier)
+	if err != nil {
+		cm, err = k8scm.createMetadataCM(identifier)
+		if err != nil {
+			if strings.Contains(err.Error(), "already exists") {
```

Comment:
The correct way to check for various kinds of errors is with `apierrors` pkg, see this for example https://github.com/kubernetes/kubernetes/blob/7f23a743e8c23ac6489340bbb34fa6f1d392db9d/test/utils/create_resources.go#L191

Reply:
Also, existing CM is not an error